gobblin-metastore/src/main/java/gobblin/metastore/nameParser/GuidDatasetUrnStateStoreNameParser.java

```diff
@@ -0,0 +1,158 @@
+import com.google.common.io.LineReader;
+
+import gobblin.util.guid.Guid;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
+
+/**
+ * Implements {@link DatasetUrnStateStoreNameParser} using {@link Guid}.
+ */
+public class GuidDatasetUrnStateStoreNameParser implements DatasetUrnStateStoreNameParser {
+  private static final String TMP_SUFFIX = "_tmp";
+
+  private final FileSystem fs;
+  private Path versionIdentifier;
+
+  @AllArgsConstructor
+  enum StateStoreNameVersion {
+    /**
+     * DatasetUrn is directly used as the state store name.
+     * This is the initial status when {@link GuidDatasetUrnStateStoreNameParser} is first enabled.
+     * It will be migrated to {@link StateStoreNameVersion#V1}.
+     */
+    V0(StringUtils.EMPTY),
+    /**
+     * DatasetUrn is hashed into shorten name.
+     */
+    V1("datasetUrnNameMapV1");
+
+    @Getter
+    private String datasetUrnNameMapFile;
+  }
+
+  private final StateStoreNameVersion version;
+  @VisibleForTesting
+  protected final BiMap<String, String> sanitizedNameToDatasetURNMap;
+
+  public GuidDatasetUrnStateStoreNameParser(FileSystem fs, Path jobStatestoreRootDir)
+      throws IOException {
+    this.fs = fs;
+    this.sanitizedNameToDatasetURNMap = Maps.synchronizedBiMap(HashBiMap.<String, String>create());
+    this.versionIdentifier = new Path(jobStatestoreRootDir, StateStoreNameVersion.V1.getDatasetUrnNameMapFile());
+    if (this.fs.exists(versionIdentifier)) {
+      this.version = StateStoreNameVersion.V1;
+      try (InputStream in = this.fs.open(versionIdentifier)) {
+        LineReader lineReader = new LineReader(new InputStreamReader(in, Charsets.UTF_8));
+        String shortenName = lineReader.readLine();
+        while (shortenName != null) {
+          String datasetUrn = lineReader.readLine();
+          this.sanitizedNameToDatasetURNMap.put(shortenName, datasetUrn);
+          shortenName = lineReader.readLine();
+        }
+      }
+    } else {
+      this.version = StateStoreNameVersion.V0;
+    }
+  }
+
+  /**
+   * Get datasetUrn for the given state store name.
+   * If the state store name can be found in {@link #sanitizedNameToDatasetURNMap}, the original datasetUrn will be returned.
+   * Otherwise, the state store name will be returned.
+   */
+  @Override
+  public String getDatasetUrnFromStateStoreName(String stateStoreName)
+      throws IOException {
+    if (version == StateStoreNameVersion.V0 || !this.sanitizedNameToDatasetURNMap.containsKey(stateStoreName)) {
+      return stateStoreName;
+    } else {
+      return this.sanitizedNameToDatasetURNMap.get(stateStoreName);
+    }
+  }
+
+  /**
+   * Merge the given {@link Collection} of datasetUrns with {@link #sanitizedNameToDatasetURNMap}, and write the results
+   * to the {@link Path} of {@link #versionIdentifier}, which is a text file.
+   */
+  @Override
+  public void persistDatasetUrns(Collection<String> datasetUrns)
+      throws IOException {
```

Comment:
We need to merge with the old mapping because one job execution may not process workunits from all datasets. The parameter `Collection<String> datasetUrns` only contains datasetUrns that are processed in the current job execution.